File: is_string_palindrome.py
def is_palindrome(string: str) -> bool:
    cleaned_string = "".join(c.lower() for c in string if c.isalnum())
    left, right = 0, len(cleaned_string) - 1
    while left < right:
        if cleaned_string[left] != cleaned_string[right]:
            return False
        left += 1
        right -= 1
    return True


def check(label, actual, expected):
    ok = actual == expected
    print(f"{'PASS' if ok else 'FAIL'}  {label}" + ("" if ok else f"  -> expected {expected!r}, got {actual!r}"))
    return ok


if __name__ == "__main__":
    results = [
        check("is_palindrome('A man, a plan, a canal: Panama')", is_palindrome('A man, a plan, a canal: Panama'), True),
        check("is_palindrome('race a car')", is_palindrome('race a car'), False),
        check("is_palindrome('')", is_palindrome(''), True),
        check('is_palindrome("No \'x\' in Nixon")', is_palindrome("No 'x' in Nixon"), True),
        check("is_palindrome('12321')", is_palindrome('12321'), True),
        check("is_palindrome('12345')", is_palindrome('12345'), False),
    ]
    print(f"{sum(results)}/{len(results)} passed")

File: IsStringPalindrome.java
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class IsStringPalindrome {

    public static void main(String[] args) {
        check("isPalindrome(\"A man, a plan, a canal: Panama\")", isPalindrome("A man, a plan, a canal: Panama"), true);
        check("isPalindrome(\"race a car\")", isPalindrome("race a car"), false);
        check("isPalindrome(\"\")", isPalindrome(""), true);
        check("isPalindrome(\"No 'x' in Nixon\")", isPalindrome("No 'x' in Nixon"), true);
        check("isPalindrome(\"12321\")", isPalindrome("12321"), true);
        check("isPalindrome(\"12345\")", isPalindrome("12345"), false);
        System.out.println(passed + "/" + total + " passed");
    }

    private static int passed = 0, total = 0;

    private static void check(String label, Object actual, Object expected) {
        total++;
        boolean ok = Objects.deepEquals(actual, expected);
        if (ok) {
            passed++;
        }
        System.out.println(
                (ok ? "PASS  " : "FAIL  ") + label + (ok ? "" : "  -> expected " + show(expected) + ", got " + show(actual)));
    }

    private static String show(Object value) {
        String text = Arrays.deepToString(new Object[] {value});
        return text.substring(1, text.length() - 1);
    }

    public static boolean isPalindrome(String string) {
        var cleaned = new StringBuilder();
        for (int i = 0; i < string.length(); i++) {
            char c = string.charAt(i);
            if (Character.isLetterOrDigit(c)) {
                cleaned.append(Character.toLowerCase(c));
            }
        }
        int left = 0;
        int right = cleaned.length() - 1;
        while (left < right) {
            if (cleaned.charAt(left) != cleaned.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

}
